src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -1205,91 +1212,121 @@ and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
-        | None -> S.new_bv (Some e1.pos) c1.res_typ
-        | Some _ ->
-          (match (e1 |> U.unascribe |> SS.compress).n with
-           | Tm_name scrutinee_bv -> scrutinee_bv
-           | _ ->
-             raise_error (Errors.Fatal_UnexpectedTerm,
-               "The scrutinee must be a variable when a return annotation is supplied with a match") e1.pos) in
-      guard_x,
-      eqns |> List.map (tc_eqn guard_x env_branches ret_opt) in
+        let b, asc =
+          let bs, asc = SS.open_ascription [b] asc in
+          let b = List.hd bs in
+          //we set the sort of the binder to be the type of e1
+          {b with binder_bv={b.binder_bv with sort=c1.res_typ}}, asc in
+        //b is in scope for asc
+        let env_asc = Env.push_binders env [b] in
+        let asc, g_asc =
+          match asc with
+          | Inl t, None ->
+            let k, _ = U.type_u () in
+            let t, _, g = tc_check_tot_or_gtot_term env_asc t k "" in
+            (Inl t, None), g
+          | Inr c, None ->
+            let c, _, g = tc_comp env_asc c in
+            (Inr c, None), g
+          | _ -> 
+            raise_error (Errors.Fatal_UnexpectedTerm,
+              "Tactic is not yet supported with match returns") (Env.get_range env) in
+
+        //we need to close g_asc with the binder b
+        env,
+        Some (b, asc),
+        Env.conj_guard g1 (Env.close_guard env_asc [b] g_asc) in
+
+    //g1 is now the guard for the scrutinee and the ascription
+    //  and it is well-formed in env
+
+    //the logical variable for the scrutinee
+    let guard_x = S.new_bv (Some e1.pos) c1.res_typ in
+    let t_eqns = eqns |> List.map (tc_eqn guard_x env_branches ret_opt) in
 
     let c_branches, g_branches, erasable =
-      let cases, g, erasable =
-          List.fold_right
-                (fun (branch, f, eff_label, cflags, c, g, erasable_branch) (caccum, gaccum, erasable) ->
-                     (f, eff_label, cflags, c)::caccum,
-                     Env.conj_guard g gaccum,
-                     erasable || erasable_branch)
-                t_eqns
-                ([], Env.trivial_guard, false) in
-
-      (* note that bind_cases adds an exhaustiveness check *)
-      match ret_or_res_typ with
-      | Inr res_t ->
-        //no return annotation, just bind_cases
-        TcUtil.bind_cases env res_t cases guard_x, g, erasable
-
-      | Inl (Inl t) ->  //a return annotation, with type
-        //set the type in the lcomp of the branches, and then bind_cases
-
-        let cases = List.map
-          (fun (f, eff_label, cflags, c) ->
-             (f, eff_label, cflags, (fun b -> TcComm.set_result_typ_lc (c b) t))) cases in
+      match ret_opt with
+      | Some (b, (Inr c, _)) ->  //a return annotation, with computation type
 
-        TcUtil.bind_cases env t cases guard_x, g, erasable
+        //c has b free, so substitute it with the scrutinee
+        let c = SS.subst_comp [NT (b.binder_bv, e1)] c in
```

Comment:
This is where we compute the type of the `match` result by substituting binder with `e1`.